dask_cuda/proxy_object.py

```diff
@@ -0,0 +1,483 @@
+                self._obj_pxy["obj"] = distributed.protocol.serialize(
+                    self._obj_pxy["obj"], serializers
+                )
+                self._obj_pxy["serializers"] = serializers
+
+            assert serializers == self._obj_pxy["serializers"]
+            return self._obj_pxy["obj"]
+
+    def _obj_pxy_deserialize(self):
+        """Inplace deserialization of the proxied object
+
+        Returns
+        -------
+        ret : object
+            The proxied object (deserialized)
+        """
+        with self._obj_pxy_lock:
+            if self._obj_pxy["serializers"] is not None:
+                header, frames = self._obj_pxy["obj"]
+                self._obj_pxy["obj"] = distributed.protocol.deserialize(header, frames)
+                self._obj_pxy["serializers"] = None
+            return self._obj_pxy["obj"]
+
+    def _obj_pxy_is_cuda_object(self):
+        """Inplace deserialization of the proxied object
+
+        Returns
+        -------
+        ret : object
+            The proxied object (deserialized)
+        """
+        with self._obj_pxy_lock:
+            return self._obj_pxy["is_cuda_object"]
+
+    def __getattr__(self, name):
+        with self._obj_pxy_lock:
+            typename = self._obj_pxy["typename"]
+            if name in _FIXED_ATTRS:
+                try:
+                    return self._obj_pxy["fixed_attr"][name]
+                except KeyError:
+                    raise AttributeError(
+                        f"type object '{typename}' has no attribute '{name}'"
+                    )
+
+            return getattr(self._obj_pxy_deserialize(), name)
+
+    def __str__(self):
+        return str(self._obj_pxy_deserialize())
+
+    def __repr__(self):
+        with self._obj_pxy_lock:
+            typename = self._obj_pxy["typename"]
+            ret = (
+                f"<{dask.utils.typename(type(self))} at {hex(id(self))} for {typename}"
+            )
+            if self._obj_pxy["serializers"] is not None:
+                ret += f" (serialized={repr(self._obj_pxy['serializers'])})>"
+            else:
+                ret += f" at {hex(id(self._obj_pxy['obj']))}>"
+            return ret
+
+    @property
+    def __class__(self):
+        with self._obj_pxy_lock:
+            try:
+                return self.__obj_pxy_cache["type_serialized"]
+            except KeyError:
+                ret = pickle.loads(self._obj_pxy["type_serialized"])
+                self.__obj_pxy_cache["type_serialized"] = ret
+                return ret
+
+    def __sizeof__(self):
+        with self._obj_pxy_lock:
+            if self._obj_pxy["serializers"] is not None:
+                frames = self._obj_pxy["obj"][1]
+                return sum(map(distributed.utils.nbytes, frames))
+            else:
+                return sys.getsizeof(self._obj_pxy_deserialize())
+
+    def __len__(self):
+        return len(self._obj_pxy_deserialize())
```

Comment:
Yes currently `len(x)` will deserialize `x`. What you suggest, we are already doing with `x.name`, which I added because worker will access `x.name` before execution tasks. 
My plan is to address this issue in a following PR when we have some more experience using JIT deserialization. I suspect we will need to handle a range of attributes.

Reply:
Makes sense, thanks for the continued work on it.